server/org.eclipse.emfcloud.ecore.glsp/src/main/java/org/eclipse/emfcloud/ecore/glsp/operationhandler/CreateEcoreEdgeOperationHandler.java

```diff
@@ -52,56 +53,76 @@ public void executeOperation(CreateEdgeOperation operation, GModelState modelSta
-
-			if (elementTypeId.equals(Types.BIDIRECTIONAL_REFERENCE)
-					|| elementTypeId.equals(Types.BIDIRECTIONAL_COMPOSITION)) {
-				EReference opposite = createReference(targetEClass, sourceEclass, elementTypeId);
-				modelAccess.addEReference(EcoreModelState.getModelState(modelState), opposite, targetEClass);
-				
-				modelAccess.setOpposite(EcoreModelState.getModelState(modelState), reference, opposite);
-				modelAccess.setOpposite(EcoreModelState.getModelState(modelState), opposite, reference);
-				
-				if (elementTypeId.equals(Types.BIDIRECTIONAL_REFERENCE)) {
-					NotationElement sourceNotationElement = modelIndex.getNotation(sourceEclass).get();
-					NotationElement targeNotationElement = modelIndex.getNotation(targetEClass).get();
-
-					for (NotationElement element : diagram.getElements()) {
-						if (element.equals(sourceNotationElement)) {
-							break;
-						}
-						if (element.equals(targeNotationElement)) {
-							reference = reference.getEOpposite();
-							break;
-						}
-					}
-				}
-			}
+			EGenericType genericType = createGenericType(sourceEClass, targetEClass);
+			Edge inheritanceEdge = createInheritanceEdge(genericType);
+			modelAccess.addESuperType(EcoreModelState.getModelState(modelState), targetEClass, sourceEClass,
+					inheritanceEdge);
+		} else if (elementTypeId.equals(Types.REFERENCE) || elementTypeId.equals(Types.COMPOSITION)) {
+			EReference reference = createReference(sourceEClass, targetEClass, elementTypeId);
+			Edge referenceEdge = createEdge(sourceEClass, reference);
+			modelAccess.addEReference(EcoreModelState.getModelState(modelState), reference, sourceEClass,
+					referenceEdge);
+		} else if (elementTypeId.equals(Types.BIDIRECTIONAL_REFERENCE)
+				|| elementTypeId.equals(Types.BIDIRECTIONAL_COMPOSITION)) {
+
+			EReference reference = createReference(sourceEClass, targetEClass, elementTypeId);
+			Edge referenceEdge = createEdge(sourceEClass, reference);
+			EReference opposite = createReference(targetEClass, sourceEClass, elementTypeId);
+			Edge oppositeEdge = createEdge(targetEClass, opposite);
+
+			modelAccess.addEReferenceBidirectional(EcoreModelState.getModelState(modelState), reference, opposite,
+					sourceEClass, targetEClass, referenceEdge, oppositeEdge,
+					elementTypeId.equals(Types.BIDIRECTIONAL_REFERENCE));
 		}
 	}
 
-	private EReference createReference(EClass source, EClass target, String elementTypeId) {
+	protected String getSemanticProxyUri(EGenericType genericType) {
+		return EcoreUtil.getURI(genericType).fragment();
+	}
+
+	protected EGenericType createGenericType(EClass sourceClass, EClass superClass) {
+		EGenericType genericType = EcoreFactory.eINSTANCE.createEGenericType();
+		genericType.setEClassifier(superClass);
+		sourceClass.getEGenericSuperTypes().add(genericType);
+		return genericType;
+	}
+	
+	protected Edge createInheritanceEdge(EGenericType genericType) {
```

Comment:
Same comment as above; inheritance edges and generic types are two different concepts